### PR TITLE
feat(core): Enforce config file permissions on startup

### PR DESCRIPTION
**Pull request by tomi**
## Summary

Config file contains the instance encryption key, which should be treated as a secret. We don't want it to be accessible by all users, so this PR changes the behaviour to make it readable and writable only by the owner of the file (i.e. the same user which the n8n application is running as) if `N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS` is set to `true`. Otherwise a warning will be logged, unless the env var is explicitly set to `false`. This behaviour is similar to what `ssh` does with private keys, i.e. it refuses to use the key unless it has limited permissions.

### 1. Settings file doesn't exists, no env var set

```
➜ ls -ld ~/.n8n-test/.n8n/config
ls: /Users/tomi/.n8n-test/.n8n/config: No such file or directory

➜ N8N_USER_FOLDER=~/.n8n-test pn start

No encryption key found - Auto-generated and saved to: /Users/tomi/.n8n-test/.n8n/config
Permissions 0644 for n8n settings file /Users/tomi/.n8n-test/.n8n/config are too wide. This is ignored for now, but in the future n8n will attempt to change the permissions automatically. To automatically enforce correct permissions now set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=true, or turn this check off set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.
Initializing n8n process
n8n ready on 0.0.0.0, port 5678

➜ ls -ld ~/.n8n-test/.n8n/config
-rw-r--r--@ 1 tomi  staff  56 Oct 22 16:03 /Users/tomi/.n8n-test/.n8n/config
```

### 2. Settings file exists, incorrect permissions, env var set to `false`

```
➜ ls -ld ~/.n8n-test/.n8n/config
-rw-r--r--@ 1 tomi  staff  56 Oct 22 16:03 /Users/tomi/.n8n-test/.n8n/config

➜ N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false N8N_USER_FOLDER=~/.n8n-test pn start

User settings loaded from: /Users/tomi/.n8n-test/.n8n/config
Initializing n8n process
n8n ready on 0.0.0.0, port 5678
Version: 1.64.0
```

### 3. Settings file exists, incorrect permissions, env var set to `true`

```
➜ ls -ld ~/.n8n-test/.n8n/config
-rw-r--r--@ 1 tomi  staff  56 Oct 22 16:03 /Users/tomi/.n8n-test/.n8n/config

➜ N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=true N8N_USER_FOLDER=~/.n8n-test pn start

Permissions 0644 for n8n settings file /Users/tomi/.n8n-test/.n8n/config are too wide. Changing permissions to 0600..
User settings loaded from: /Users/tomi/.n8n-test/.n8n/config
Initializing n8n process
n8n ready on 0.0.0.0, port 5678
Version: 1.64.0

❯ ls -ld ~/.n8n-test/.n8n/config
-rw-------@ 1 tomi  staff  56 Oct 22 16:03 /Users/tomi/.n8n-test/.n8n/config
```

### 4. Settings file exists, correct permissions, env var not set

```
❯ ls -ld ~/.n8n-test/.n8n/config
-rw-------@ 1 tomi  staff  56 Oct 22 16:03 /Users/tomi/.n8n-test/.n8n/config

➜ N8N_USER_FOLDER=~/.n8n-test pn start

User settings loaded from: /Users/tomi/.n8n-test/.n8n/config
Initializing n8n process
n8n ready on 0.0.0.0, port 5678
Version: 1.64.0
```

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/PAY-2125/some-way-to-chmod-the-n8n-folder-config-file

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

**Review by valya — DISMISSED** [review]

File: packages/core/src/InstanceSettings.ts [hunk 1]
```diff
@@ -170,4 +172,31 @@ export class InstanceSettings {
 		this.settings = settings;
 		writeFileSync(this.settingsFile, JSON.stringify(settings, null, '\t'), 'utf-8');
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/core/src/InstanceSettings.ts:173]: this code is called for newly created files as well. That's what I was referring to.
we can start setting the mode on newly created files, and that way the config files in tests don't need to skip the permission check.

File: packages/core/src/InstanceSettings.ts [hunk 2]
```diff
@@ -168,6 +175,93 @@ export class InstanceSettings {
...
 	private save(settings: Settings) {
 		this.settings = settings;
-		writeFileSync(this.settingsFile, JSON.stringify(settings, null, '\t'), 'utf-8');
+		writeFileSync(this.settingsFile, JSON.stringify(this.settings, null, '\t'), {
+			mode: this.enforceSettingsFilePermissions.enforce ? 0o600 : undefined,
+			encoding: 'utf-8',
+		});
+	}
+
+	private loadEnforeSettingsFilePermissionsFlag(): {
```
Comment (netroy) [inline packages/core/src/InstanceSettings.ts:184]:
```suggestion
	private loadEnforceSettingsFilePermissionsFlag(): {
```

File: packages/core/src/InstanceSettings.ts [hunk 3]
```diff
@@ -168,6 +175,93 @@ export class InstanceSettings {
...
+		isSet: boolean;
+		enforce: boolean;
+	} {
+		const envVar = process.env.N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS?.toLowerCase();
+		const isEnvVarSet = !!envVar;
```
Comment (netroy) [inline packages/core/src/InstanceSettings.ts:189]: since `!!('true')` and `!!('false')` will both evaluate to `true`, we should change this to:

```suggestion
		const isEnvVarSet = envVar === 'true';
```

Even better option would be to create a `@Config` class instead, and benefit from the already existing (and tested) boolean config parsing code.

**Review by netroy — CHANGES_REQUESTED** [review]

File: packages/core/src/InstanceSettings.ts (hunk 3 above)
Comment (netroy) [reply packages/core/src/InstanceSettings.ts:189]: > since `core` doesn't have a dependency to `@n8n/config`

One of the primary purposes of `@n8n/config` was to extract out all of backend config, so that all backend specific packages could use them. So, adding this dependency would be perfectly fine.

Comment (tomi) [reply packages/core/src/InstanceSettings.ts]: 👍 done

File: packages/@n8n/config/src/index.ts [hunk 4]
```diff
@@ -97,4 +99,7 @@ export class GlobalConfig {
 
 	@Nested
 	multiMainSetup: MultiMainSetupConfig;
+
+	@Nested
+	security: SecurityConfig;
```
Comment (netroy) [inline packages/@n8n/config/src/index.ts:108]: we don't need to add everything in the global config. we can move the config class to core as well.

File: packages/core/src/InstanceSettings.ts [hunk 5]
```diff
@@ -185,8 +183,8 @@ export class InstanceSettings {
 		isSet: boolean;
 		enforce: boolean;
 	} {
-		const envVar = process.env.N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS?.toLowerCase();
-		const isEnvVarSet = !!envVar;
+		const { enforceSettingsFilePermissions } = Container.get(SecurityConfig);
```
Comment (netroy) [inline packages/core/src/InstanceSettings.ts:186]: we can replace the `Container.get` with a constructor injection.

**Review by netroy — CHANGES_REQUESTED** [review]

File: packages/@n8n/config/src/index.ts (hunk 4 above)
Comment (tomi) [reply packages/@n8n/config/src/index.ts]: Moved it to core. But wouldn't it make sense to have all the app configuration in a single location instead of being spread around the codebase?

File: packages/core/src/InstanceSettingsConfig.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { Config, Env } from '@n8n/config';
+
+@Config
+export class InstanceSettingsConfig {
```
Comment (tomi) [inline packages/core/src/InstanceSettingsConfig.ts:4]: Not maybe the best name. Thought using `SecurityConfig` but if we move the other envs `InstanceSettings` uses (`N8N_USER_FOLDER` and `N8N_ENCRYPTION_KEY`) that name wouldn't make sense

File: packages/cli/src/executions/__tests__/execution-recovery.service.test.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import { stringify } from 'flatted';
 import { mock } from 'jest-mock-extended';
 import { InstanceSettings } from 'n8n-core';
+import { InstanceSettingsConfig } from 'n8n-core/src/InstanceSettingsConfig';
```
Comment (netroy) [inline packages/cli/src/executions/__tests__/execution-recovery.service.test.ts:4]:
```suggestion
```

File: packages/cli/src/executions/__tests__/execution-recovery.service.test.ts [hunk 8]
```diff
@@ -22,7 +23,7 @@ import { setupMessages } from './utils';
 
 describe('ExecutionRecoveryService', () => {
 	const push = mockInstance(Push);
-	const instanceSettings = new InstanceSettings();
+	const instanceSettings = new InstanceSettings(new InstanceSettingsConfig());
```
Comment (netroy) [inline packages/cli/src/executions/__tests__/execution-recovery.service.test.ts:26]:
```suggestion
	const instanceSettings = new InstanceSettings(mock());
```

File: packages/cli/test/integration/pruning.service.test.ts [hunk 9]
```diff
@@ -1,5 +1,6 @@
 import { mock } from 'jest-mock-extended';
 import { BinaryDataService, InstanceSettings } from 'n8n-core';
+import { InstanceSettingsConfig } from 'n8n-core/src/InstanceSettingsConfig';
```
Comment (netroy) [inline packages/cli/test/integration/pruning.service.test.ts:3]:
```suggestion
```

File: packages/cli/test/integration/pruning.service.test.ts [hunk 10]
```diff
@@ -22,7 +23,7 @@ import { mockInstance } from '../shared/mocking';
 
 describe('softDeleteOnPruningCycle()', () => {
 	let pruningService: PruningService;
-	const instanceSettings = new InstanceSettings();
+	const instanceSettings = new InstanceSettings(new InstanceSettingsConfig());
```
Comment (netroy) [inline packages/cli/test/integration/pruning.service.test.ts:26]:
```suggestion
	const instanceSettings = new InstanceSettings(mock());
```

File: packages/workflow/src/result.ts [hunk 11]
```diff
@@ -11,3 +13,18 @@ export const createResultError = <E = unknown>(error: E): ResultError<E> => ({
...
 	error,
 });
+
+/**
+ * Executes the given function and converts it to a Result object.
+ *
+ * @example
+ * const result = toResult(() => fs.writeFileSync('file.txt', 'Hello, World!'));
+ */
+export const toResult = <T, E extends Error = Error>(fn: () => T): Result<T, E> => {
```
Comment (netroy) [inline packages/workflow/src/result.ts:23]: nice. I did not realize that we had already added a `Result` type. Thanks.
but, since this is not used in editor or nodes (and since we don't know if those teams want to use this), can we please move this to core as well. We can do that in another PR though.
Also, we should add tests for all this code.

**Review by netroy — CHANGES_REQUESTED** [review]
we really need something like `mockInstance` for unit tests that instead of creating actual instances via typedi, creates mocks for all dependencies except the ones passed in as a second argument. 
That way we can avoid loading up any additional code in unit tests.

**Review by netroy — APPROVED** [review]
Thanks 💟

**Comment by janober** [comment]
Got released with [`n8n@1.65.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.65.0)
